gluten-data/src/main/scala/org/apache/spark/sql/execution/ColumnarWriteFilesExec.scala

```diff
@@ -0,0 +1,113 @@
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.connector.write.WriterCommitMessage
+import org.apache.spark.sql.execution.datasources.{BasicWriteTaskStats, ExecutedWriteSummary, FileFormat, PartitioningUtils, WriteFilesExec, WriteFilesSpec, WriteTaskResult}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+import shaded.parquet.com.fasterxml.jackson.databind.ObjectMapper
+
+import scala.collection.mutable
+
+class ColumnarWriteFilesExec(
+    child: SparkPlan,
+    fileFormat: FileFormat,
+    partitionColumns: Seq[Attribute],
+    bucketSpec: Option[BucketSpec],
+    options: Map[String, String],
+    staticPartitions: TablePartitionSpec)
+  extends WriteFilesExec(
+    child,
+    fileFormat,
+    partitionColumns,
+    bucketSpec,
+    options,
+    staticPartitions) {
+
+  override def supportsColumnar(): Boolean = true
+
+  override def doExecuteWrite(writeFilesSpec: WriteFilesSpec): RDD[WriterCommitMessage] = {
+    assert(child.supportsColumnar)
+
+    child.session.sparkContext.setLocalProperty("writePath", writeFilesSpec.description.path)
+    child.executeColumnar().map {
+      cb =>
+        val loadedCb = ColumnarBatches.ensureLoaded(ArrowBufferAllocators.contextInstance, cb)
+        val numRows = loadedCb.column(0).getLong(0)
+
+        var updatedPartitions = Set.empty[String]
+
+        val addedAbsPathFiles: mutable.Map[String, String] = mutable.Map[String, String]()
+
+        for (i <- 0 until loadedCb.numRows() - 1) {
+          val fragments = loadedCb.column(1).getUTF8String(i + 1)
+          val objectMapper = new ObjectMapper()
+          val jsonObject = objectMapper.readTree(fragments.toString)
+          if (jsonObject.get("name").textValue().nonEmpty) {
+            updatedPartitions += jsonObject.get("name").textValue()
+          }
+          val fileWriteInfos = jsonObject.get("fileWriteInfos").elements()
+          if (updatedPartitions.size > 0 && jsonObject.get("fileWriteInfos").elements().hasNext) {
+            val writeInfo = fileWriteInfos.next();
+            val fileSize = writeInfo.get("fileSize")
+            val targetFileName = writeInfo.get("targetFileName").textValue()
+            val partitionDir = jsonObject.get("name").textValue()
+            val tmpOutputPath =
+              writeFilesSpec.description.path + "/" + partitionDir + "/" + targetFileName
+            val absOutputPathObject =
+              writeFilesSpec.description.customPartitionLocations.get(
+                PartitioningUtils.parsePathFragment(partitionDir))
+            if (absOutputPathObject.nonEmpty) {
+              val absOutputPath = absOutputPathObject.get + "/" + targetFileName
+              addedAbsPathFiles(tmpOutputPath) = absOutputPath
+            }
+          }
+        }
+
+        // TODO: need to get the partition Internal row ? numFiles, numBytes from cb.
+        val stats = BasicWriteTaskStats(Seq.empty, 0, 0, numRows)
```

Comment:
we can get metrics from `WrilteFilesTransformer`. `WrilteFilesTransformer` is belong to the previous whole stage transformer, so the metrics should be updated when we are going to `ColumnarWriteFilesExec`

Reply:
Good suggestion. Can we incorporate this change in a subsequent PR that update metrics for the `WriteTransformer` https://github.com/oap-project/gluten/pull/3690#discussion_r1432189196?

Reply:
sure